sources/bloodhound/src/main.rs

```
@@ -0,0 +1,210 @@
+use std::fs::{DirEntry, File};
+use std::io::{stdout, Error, Write};
+use std::os::unix::fs::PermissionsExt;
+use std::process::{Command, Output};
+use std::{fs, path::PathBuf};
+
+// Define some exit codes for error conditions
+const CHECKER_DISCOVERY_ERROR: i32 = 1;
+const REPORT_OUTPUT_ERROR: i32 = 2;
+const NO_CHECKS_RUN_ERROR: i32 = 3;
+
+#[derive(ValueEnum, Clone, Debug)]
+enum Format {
+    Text,
+    Json,
+}
+
+/// Command line arguments for the bloodhound program.
+#[derive(Parser, Debug)]
+struct Arguments {
+    /// Path to the directory containing checker binaries [default: /usr/libexec/bloodhound-checks]
+    #[arg(short, long = "checks")]
+    check_dir: Option<PathBuf>,
+    /// Format of the output
+    #[clap(short, long, value_enum, default_value_t=Format::Text)]
+    format: Format,
+    /// The CIS benchmark compliance level to check
+    #[arg(short, long, default_value_t = 1)]
+    level: u8,
+    /// Write output to a file at given path [default: stdout]
+    #[arg(short, long)]
+    output: Option<PathBuf>,
+}
+
+/// Discover all executable checker files for the given directory.
+/// By looking at the provided path, this does some basic checks and filtering
+/// to find the executable files that appear to be bloodhound checkers.
+/// It will parse the checker metadata to provide a mapping between a filesystem
+/// path and a specific check, including only those checkers that are within the
+/// request compliance level.
+fn find_checkers(check_dir: &PathBuf, level: u8) -> HashMap<String, CheckerMetadata> {
+    let entries: Vec<DirEntry> = fs::read_dir(check_dir)
+        .unwrap()
+        .filter_map(|file| file.ok())
+        .collect();
+    let mut result = HashMap::with_capacity(entries.len());
+
+    for entry in entries {
+        if let Ok(file_metadata) = fs::metadata(entry.path()) {
+            // Filter out any subdirectories
+            if file_metadata.is_dir() {
+                continue;
+            }
```

I feel doubtful about this check.

If we, the core distro developers, own the structure of this directory, we know we won't put directories in the checkers patch. However, OOTB users may do what they please and organize their checks differently e.g., in directories; then this check will ignore their checks without even providing a heads-up saying “FYI, ignoring unexpected directory”.

Thus, I think we either traverse directories that are placed here, or we print a message saying that the directories will be ignored, and document in OOTB that for this feature, `bloodhound` expects actual binaries.

---

This is very much intentional. We don't want someone to accidentally point `--checks` at something like `/` by mistake and have it recurse through the entire filesystem trying to run every binary it finds. The design calls out that for someone to provide checks, it needs to be a binary in a specific directory.

If OOTB users want to organize their checkers in a different structure for whatever reason, they could easily provide symlinks for which ones they want to include.

We also don't want to emit warning output on this, because at least in some cases the output is something that could be ingested into some other parsing. Emitting warnings for something like this could cause problems doing that.

It is also intentional that subdirectories are ignored. Some future work being looked at would rely on this for handling different scenarios, so it could be common even in Bottlerocket that there would be a directory that should just be silently ignored.